runtime/cudaq/domains/chemistry/uccsd.h

```diff
@@ -13,192 +13,621 @@
 #include "cudaq/utils/cudaq_utils.h"
 
 namespace cudaq {
-using SingleIndices = std::vector<std::size_t>;
-using DoubleIndices = std::vector<std::size_t>;
-using Excitations =
-    std::tuple<std::vector<SingleIndices>, std::vector<DoubleIndices>>;
-
-Excitations generateExcitations(std::size_t nElectrons, std::size_t nOrbitals) {
-  std::vector<double> sz(nOrbitals);
-  for (auto &i : cudaq::range(nOrbitals))
-    sz[i] = (i % 2 == 0) ? 0.5 : -0.5;
-
-  std::vector<SingleIndices> singles;
-  for (auto &r : cudaq::range(nElectrons))
-    for (std::size_t p = nElectrons; p < nOrbitals; p++) {
-      if (sz[p] - sz[r] == 0)
-        singles.emplace_back(SingleIndices{r, p});
-    }
-
-  std::vector<DoubleIndices> doubles;
-  for (auto &s : cudaq::range(nElectrons - 1))
-    for (std::size_t r = s + 1; r < nElectrons; r++)
-      for (std::size_t q = nElectrons; q < nOrbitals - 1; q++)
-        for (std::size_t p = q + 1; p < nOrbitals; p++)
-          if (sz[p] + sz[q] - sz[r] - sz[s] == 0)
-            doubles.emplace_back(DoubleIndices{s, r, q, p});
-
-  return std::make_tuple(singles, doubles);
-}
 
-template <typename KernelBuilder>
-void singletExcitation(KernelBuilder &&kernel, QuakeValue &qubits,
-                       QuakeValue &theta, const SingleIndices &indices) {
-  auto r = indices.front();
-  auto p = indices.back();
-  kernel.exp_pauli(0.5 * theta, "YX", qubits[r], qubits[p]);
-  kernel.exp_pauli(-0.5 * theta, "XY", qubits[r], qubits[p]);
+/// @brief An excitation_list is a vector of lists of indices.
+using excitation_list = std::vector<std::vector<std::size_t>>;
+
+/// @brief UCCSD excitations for single and double excitations.
+struct excitations {
+  excitation_list singles_alpha;
+  excitation_list singles_beta;
+  excitation_list doubles_mixed;
+  excitation_list doubles_alpha;
+  excitation_list doubles_beta;
```

Comment:
Question only: is there a reason to split them out explicitly vs. just having `singles` and `doubles` lists?
For example, we need to write 3 identical for loops to handle double excitations.

Reply:
They are not identical loops. But we can concatenate singles_alpha and singles_beta in singles and doubles_mixed, doubles_alpha, and doubles_beta into doubles. The reason I did this splitting because I want to further optimize the UCCSD quantum circuit for the restricted spin case. In restricted spin case,  alpha and beta electrons are identical and therefore singles_alpha equals singles_beta and doubles_alpha equals doubles_beta. However, this is not possible when you have the number of alpha electrons does not equal the number of beta electrons.